### ISSUE TITLE
writing vtu file: Error `list index out of range`

### ISSUE THREAD

**Issue by ArcticSnow** [Chrismarsh/mesher#20]
Hi, 
I find your library really interesting. I was able to install it within a fresh conda VE with pip (Python 3.8 on Ubuntu 20.04) after some battling with dependencies. Trying to apply the library on a DEM, I encountered the following error:

```bash
(mesher) simonfi@gulogulo:~/github/TopoPyScale_examples/ex1_norway_finse/inputs/dem$ python /home/simonfi/miniconda3/envs/mesher/bin/mesher.py mesher_config.py 
Using 8 CPUs and per-CPU memory cache = 3.75%
Creating output file that is 211P x 192L.
Processing ./ASTER_Finse.tif [1/1] : 0...10...20...30...40...50...60...70...80...90...100 - done.
Using FID = 0 as the largest continuous area.
Simplifying extents by buffer distance = -10
Converting polygon to linestring
/home/simonfi/miniconda3/envs/mesher/bin/mesher --poly-file /home/simonfi/github/TopoPyScale_examples/ex1_norway_finse/inputs/dem//mesher_config/ASTER_Finse/PLGSASTER_Finse.poly --tolerance 5 --raster /home/simonfi/github/TopoPyScale_examples/ex1_norway_finse/inputs/dem//mesher_config/ASTER_Finse/ASTER_Finse_projected.tif --area 99999980000001 --min-area 900 --error-metric rmse --lloyd 1 --interior-plgs-file /home/simonfi/github/TopoPyScale_examples/ex1_norway_finse/inputs/dem//mesher_config/ASTER_Finse/interior_PLGS.geojson
Reading interior PLGS file /home/simonfi/github/TopoPyScale_examples/ex1_norway_finse/inputs/dem//mesher_config/ASTER_Finse/interior_PLGS.geojson
Number of input PLGS vertices: 4
Meshing the triangulation...
Running 1 Lloyd iterations... done.
Number of vertices: 71798
Number of triangles: 142893
Meshing took 85.42 s
Reading nodes
Length of invalid nodes = 0
Reading nodes took 0.64 s
Reading in neighbour file
Reading neigh file took 0.27 s
Repairing invalid triangles if needed...
Repairing nodes took 0.33 s
Computing parameters and initial conditions
Rank 0 writing output pickle
Total time took 0.57 s
Writing vtu file...
Traceback (most recent call last):
  File "/home/simonfi/miniconda3/envs/mesher/bin/mesher.py", line 1466, in <module>
    main()
  File "/home/simonfi/miniconda3/envs/mesher/bin/mesher.py", line 759, in main
    write_vtu(base_dir + base_name + '.vtu', mesh, params, ics)
  File "/home/simonfi/miniconda3/envs/mesher/bin/mesher.py", line 1443, in write_vtu
    output = data[elem]
IndexError: list index out of range
```
My config file reads as follow:
```python
dem_filename = './ASTER_Finse.tif'

max_area= 9999999**2  #Effectively unlimited upper area -- allow tolerance check to refine it further
max_tolerance = 5    #1m max RMSE between triangle and underlying elevation
min_area = 30**2     #triangle area below which we will no longer refine, regardless of max_tolerance
errormetric = 'rmse'
lloyd_itr=1
simplify=True
MPI_nworkers=1
```
Any idea what might cause this problem

**Comment by Chrismarsh** [comment]
Can you confirm what version of mesher you are  using? 
`pip list | grep mesher`

>after some battling with dependencies

yes, no matter where and what dependencies can be tough. If you have any suggestions for the documentation please let me know.

**Comment by ArcticSnow** [comment]
It seems Pip is installing version 2.0.1

As for the install, I had one issue with some p11 dependencies. I updated the documentation process with one line:
```bash
# system wide dependencies
sudo apt-get install libgdal-dev
sudo apt-get install libcgal-dev
sudo apt-get install gdal-bin
sudo apt-get install libcgal-dev
sudo apt-get install libboost-filesystem-dev
sudo apt-get install libboost-program-options-dev
sudo apt-get install libmetis-dev

# on Ubuntu 20.04+
sudo apt-get install python3-gdal

# Create a conda VE
conda create -y --name mesher python=3.8
conda activate mesher

# install required dependencies:
conda install -c conda-forge p11-kit

pip install mesher
```

Then the other mess is gdal. Later, yesterday I installed rasterio (using conda) in the same VE, and it ruined gdal dependencies as conda installed another gdal specifically to the VE. The follow up struggle led me to find out that purging pip was needed to rebuilt wheels properly, and install a fresf VE. Now I am back to the same error.

So it seems the version pip fetches corresponds to the latest release on the github repo.

**Comment by Chrismarsh** [comment]
Sorry about that. I had an edge case after the large MPI rewrite such that when no landsurface parameters were computed, the triangle areas were not computed despite being expected. This should be resolved in [v2.0.3](https://github.com/Chrismarsh/mesher/tree/2.0.3) which is now pushed to PyPi.

I was able to reproduce your DEM and I would suggest that you set your min triangle size to be approx that of the DEM, otherwise you can get lumpy over refinement. I would also suggest coarsening the tolerance slightly (~10m) to get the full impact of the variable resolution. Otherwise, the algorithm will be forced to use a more regular sized triangle. Mesher really excels when it is approximating a surface instead of exactly representing it

```
dem_filename = './ASTER_Finse.tif'

max_area= 9999999**2  #Effectively unlimited upper area -- allow tolerance check to refine it further
max_tolerance = 10    #1m max RMSE between triangle and underlying elevation
min_area = 100**2     #triangle area below which we will no longer refine, regardless of max_tolerance
errormetric = 'rmse'
lloyd_itr=1
simplify=True
MPI_nworkers=1
```

![image](https://user-images.githubusercontent.com/630436/215550676-a5720131-a02f-4e65-b125-19fd293a5c29.png)



Thanks for the heads up on p11. I have no idea what that is or why it's needed! [But I've added it to the docs.](https://mesher-hydro.readthedocs.io/en/latest/installation.html#conda)

Agreed that gdal is tough. I don't use conda, so I'm not super familiar with it's nuances other than having gdal within the VE + system results in weird interactions in my experience. 

n.b., I see you're one of the TopoScale authors. I've always been very interested in TopoScale in the context of [CHM](https://github.com/Chrismarsh/CHM). I've just never gotten around to it incorporating your algorithms. Don't hestiate to drop me an email (chris.marsh@usask.ca) if you'd like to chat further.

**Comment by ArcticSnow** [comment]
Thank you, that works like a charm!
I'll get in touch with you directly for the rest ;)
